File: src/Oisys.Web/ClientApp/src/app/shared/models/credit-memo-line-item.ts
import { JsonModelBase } from "./json-model-base";
import { OrderLineItem } from "./order-line-item";

export class CreditMemoLineItem extends JsonModelBase {
  public id: number;
  public creditMemoId: number;
  public orderLineItemId: number;
  public itemId: number;
  public quantity: number;
  public shouldAddBackToInventory: boolean;

  private _orderLineItem: OrderLineItem;
  get orderLineItem() {
    return this._orderLineItem;
  }
  set orderLineItem(orderLineItem: OrderLineItem) {
    if (orderLineItem) {
      this._orderLineItem = orderLineItem;
      this.orderLineItemId = orderLineItem.id;
      this.itemId = orderLineItem.itemId;
    }
  }

  get unit() {
    return this._orderLineItem && this._orderLineItem.unit || '';
  }

  get categoryName() {
    return this._orderLineItem && this._orderLineItem.categoryName || '';
  }

  get unitPrice() {
    return this._orderLineItem && this._orderLineItem.unitPrice || 0;
  }

  get totalPrice() {
    return this.quantity * this.unitPrice;
  }

  constructor();
  constructor(creditMemoLineItem: CreditMemoLineItem);
  constructor(creditMemoLineItem?: any) {
    super();

    this.id = creditMemoLineItem && creditMemoLineItem.id || 0;
    this.creditMemoId = creditMemoLineItem && creditMemoLineItem.creditMemoId || 0;
    this.orderLineItemId = creditMemoLineItem && creditMemoLineItem.orderLineItemId || 0;
    this.itemId = creditMemoLineItem && creditMemoLineItem.itemId || 0;
    this.quantity = creditMemoLineItem && creditMemoLineItem.quantity || 0;
    this.shouldAddBackToInventory = creditMemoLineItem && creditMemoLineItem.shouldAddBackToInventory || false;

    this.orderLineItem = (creditMemoLineItem && creditMemoLineItem.orderLineItem) ? new OrderLineItem(creditMemoLineItem.orderLineItem) : undefined;
  }
}

File: src/Oisys.Web/Models/CreditMemoLineItem.cs
namespace OisysNew.Models
{
    /// <summary>
    /// <see cref="CreditMemoLineItem"/> class CreditMemodetail object.
    /// </summary>
    public class CreditMemoLineItem : ModelBase
    {
        /// <summary>
        /// Gets or sets the id of the parent credit memo.
        /// </summary>
        public long CreditMemoId { get; set; }

        /// <summary>
        /// Gets or sets the line item id of the order.
        /// </summary>
        public long OrderLineItemId { get; set; }

        /// <summary>
        /// Gets or sets property Item Id.
        /// </summary>
        public long ItemId { get; set; }

        /// <summary>
        /// Gets or sets the quantity.
        /// </summary>
        public int Quantity { get; set; }

        /// <summary>
        /// Gets or sets the unit price.
        /// </summary>
        public decimal UnitPrice { get; set; }

        /// <summary>
        /// Gets or sets the total price.
        /// </summary>
        public decimal TotalPrice { get; set; }

        /// <summary>
        /// Gets or sets a value indicating whether
        /// the item is returned to the inventory.
        /// </summary>
        public bool ReturnedToInventory { get; set; }

        /// <summary>
        /// Gets or sets property Order Line Item.
        /// </summary>
        public virtual OrderLineItem OrderLineItem { get; set; }

        /// <summary>
        /// Gets or sets the parent credit memo.
        /// </summary>
        public virtual CreditMemo CreditMemo { get; set; }

        /// <summary>
        /// Gets or sets the item linked to this credit memo item.
        /// </summary>
        public virtual Item Item { get; set; }

        /// <summary>
        /// Gets or sets the item history linked to this credit memo item.
        /// </summary>
        public virtual ItemHistory TransactionHistory { get; set; }
    }
}
